Add Let's Go Pikachu/Eevee and Sword/Shield data

Parnassius opened the pull request:
Closes #311 
Closes #284 
Closes #262 
Closes #210 
Closes #231

Parnassius commented on pokedex/data/csv/pokemon_evolution.csv:
```diff
@@ -396,10 +396,52 @@ id,evolved_species_id,evolution_trigger_id,trigger_item_id,minimum_level,gender_
 403,791,1,,53,,,,,,,,,,,,,,0,0
 404,792,1,,53,,,,,,,,,,,,,,0,0
 405,804,1,,,,,,,406,,,,,,,,,0,0
+406,462,3,83,,,,,,,,,,,,,,,0,0
+407,470,3,85,,,,,,,,,,,,,,,0,0
+408,471,3,885,,,,,,,,,,,,,,,0,0
+409,678,1,,25,1,,,,,,,,,,,,,0,0
+410,738,3,83,,,,,,,,,,,,,,,0,0
```
magnezone, leafeon, glaceon, vikavolt now evolve using thunder/leaf/ice stones

Parnassius replied:
I missed the fourth line yesterday, that's female meowstic. Meowstic already has two rows in this file (ids 360 and 372), in the first one the `gender_id` field is empty, in the second one is set to `2` (male). I think we should either keep the first record (since there's no `evolved_forme_id` column in this table), or the other two.

Parnassius replied:
Done. This leaves a gap in the `id` column, but I don't think it's a problem (and row 325 was already missing)

Parnassius commented on pokedex/data/csv/pokemon_stats.csv:
```diff
@@ -1191,7 +1191,7 @@ pokemon_id,stat_id,base_stat,effort
 199,3,80,0
 199,4,100,0
-199,5,110,3
+199,5,110,2
```
slowking now only yields 2 evs in spd

Parnassius commented on pokedex/data/csv/pokemon_stats.csv:
```diff
@@ -4081,9 +4081,9 @@ pokemon_id,stat_id,base_stat,effort
 681,2,50,0
-681,3,150,2
+681,3,140,2
 681,4,50,0
-681,5,150,1
+681,5,140,1
```
aegislash's defences were lowered to 140

Parnassius commented on pokedex/data/csv/pokemon_stats.csv:
```diff
@@ -4992,9 +5538,9 @@ pokemon_id,stat_id,base_stat,effort
 10025,5,81,0
 10025,6,104,2
 10026,1,60,0
-10026,2,150,2
+10026,2,140,2
 10026,3,50,0
-10026,4,150,1
+10026,4,140,1
```
aegislash-blade's attacks were lowered to 140

Parnassius commented on pokedex/data/csv/pokemon.csv:
```diff
@@ -905,61 +996,96 @@ id,identifier,species_id,height,weight,base_experience,order,is_default
+10115,marowak-alola,105,10,340,149,164,0
+10116,greninja-battle-bond,658,15,400,239,787,0
+10117,greninja-ash,658,15,400,288,788,0
+10118,zygarde-10-power-construct,718,12,335,243,858,0
+10119,zygarde-50-power-construct,718,50,3050,300,859,0
```
i changed those two identifiers (and added an additional `zygarde-10` below) to better represent the in-game data (2 "standard" formes, 2 power construct formes, 1 zygarde complete).
Both here and in `pokemon_forms.csv`

Parnassius commented on pokedex/data/csv/machines.csv:
```diff
@@ -1441,3 +1608,96 @@ machine_number,version_group_id,item_id,move_id
+197,20,1285,676
+198,20,1286,706
+199,20,1287,710
+200,20,1188,776
```
TMs and TRs are numbered from 00 to 99, here i used IDs 1-100 for the TMs, and 101-200 for the TRs, so there's a difference of one between the two numberings. For example TM05 is ID 6, and TR99 is ID 200.

Parnassius replied:
Shifted them down, it doesn't seem to cause problems using sqlite, not sure about others

Parnassius commented on pokedex/data/csv/move_meta_stat_changes.csv:
```diff
@@ -36,11 +36,11 @@ move_id,stat_id,change
 231,3,-1
 232,2,1
 242,3,-1
-246,1,1
```
Closes #210

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -869,8 +869,8 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
-922,xtranceiver--red,23,0,,
-923,xtranceiver--yellow,23,0,,
+922,xtransceiver--red,23,0,,
+923,xtransceiver--yellow,23,0,,
```
oops, good catch

magical reviewed (commented):
I reviewed the Let's Go changes. Looks good to me. Just a few comments, and only one that needs to be addressed (duplicate items).

Thanks for doing this! And thanks for updating the changelog and game_indices tables; they are often missed.

I'll review Sword/Shield next.

magical commented:
Oh, one thing i missed from Let's Go - it looks like the partner forms of Pikachu and Eevee are missing. They probably got lost when you dropped unobtainable pokemon from the rip.

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -227,14 +227,14 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
 227,dragon-fang,19,1000,70,
 228,silk-scarf,19,1000,10,
-229,up-grade,10,2000,30,
+229,upgrade,10,2000,30,
```
Identifiers are supposed to be semi-stable. That's why we have identifiers at all instead of just using the names verbatim. Can we leave up-grade, stick, and the various -wings as they were? Unless there's a good reason to change them.

Parnassius replied:
Reverted this, stick, *-wings, and adventure-rules/ilimas-normalium-z as well

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -303,98 +303,98 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
-314,tm10,37,10000,,
-315,tm11,37,3000,,
-316,tm12,37,5000,,
+305,tm01,37,40000,,
```
lol, i guess that's what happens when TMs reuse the same items every game. we really need a way to track cost changes. i noticed a bunch of other items had their cost changed too.

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -303,98 +303,98 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
+321,tm17,37,10000,,
+322,tm18,37,10000,,
+323,tm19,37,10000,,
+324,tm20,37,34464,,
```
34464 is an... oddly specific number. Bulbapedia says TM20 is actually 100,000 pokedollars. i guess cost isn't a uint16 any more.

Parnassius replied:
Nice catch, i completely missed this. I'll try to fix it in the next couple of days

Parnassius replied:
Bingo, it's now uint32. Only 7 items (tm20 and 6 other tms) need the third byte (so far nothing needs the fourth byte - and probably never will, i guess).

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -1085,3 +1085,524 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
+1321,precooked-burger,,150,10,
+1322,gigantamix,,15000,50,
+1323,wishing-chip,,0,,
+1324,rotom-bike,,0,,
```
2nd rotom-bike. there's another @ 1143

Parnassius replied:
Fixed, it's now `rotom-bile--water-mode`

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -1085,3 +1085,524 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
+1644,legendary-clue-1,,0,,
+1645,legendary-clue-2,,0,,
+1646,legendary-clue-3,,0,,
+1647,legendary-clue?,,0,,
```
identifier shouldn't contain "?"

Parnassius replied:
Fixed as `legendary-clue-question`, let me know if you prefer something else

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -1085,3 +1085,524 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
+1327,band-autograph,,0,,
+1328,sonias-book,,0,,
+1329,rotom-catalog,,0,,
+1330,★and458,,20,,
```
what are these ★ items? unused?

if they're junk, please delete. if not, identifiers should be ascii only.

Parnassius replied:
They are [Dynamax Crystals](https://bulbapedia.bulbagarden.net/wiki/Dynamax_Crystal), there are 300 of them, and they are distributed as events. So far only 6 of them have been made available. I replaced the star with a `dynamax-crystal-` prefix

magical commented on pokedex/data/csv/items.csv:
```diff
@@ -1085,3 +1085,524 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
+1634,style-card,,0,,
+1635,armor-pass,,0,,
+1636,rotom-bike,,0,,
+1637,rotom-bike,,0,,
```
more rotom bikes. merge them or give them a unique identifier (ala bike--green and bike--yellow).

Parnassius replied:
Those two are `rotom-bike--sparkling-white` and `rotom-bike--glistering-black`

magical commented on pokedex/data/csv/move_names.csv:
```diff
@@ -3,7 +3,7 @@ move_id,local_language_id,name
 1,4,拍擊
 1,5,Écras’Face
-1,6,Pfund
+1,6,Klaps
```
👏

magical commented on pokedex/data/csv/moves.csv:
```diff
@@ -9,7 +9,7 @@ id,identifier,generation_id,type_id,power,pp,accuracy,priority,target_id,damage_
 9,thunder-punch,1,13,75,15,100,0,10,2,7,10,1,1,17
 10,scratch,1,1,40,35,100,0,10,2,1,,5,1,5
-11,vice-grip,1,1,55,30,100,0,10,2,1,,5,1,5
+11,vise-grip,1,1,55,30,100,0,10,2,1,,5,1,5
```
i'm inclined to leave this as vice-grip

Parnassius replied:
Reverted

magical reviewed (commented):
First pass at Sword/Shield review. Found some issues in items.csv (wrong costs, duplicate bikes, invalid identifiers). Plus some minor comments and a request about Sylveon.

magical commented:
BTW, it looks like you mostly chose the same form identifiers as i did for the HOME sprites (https://github.com/veekun/pokedex/commit/b09929dfb636d2b0337edbfe6761496c7d435965).

I think the only differences are:
* darmanitan-galar-standard vs darmanitan-standard-galar
* darmanitan-galar-zen vs darmanitan-zen-galar
* zacian-crowned vs zacian-crowned-sword
* zamazenta-crowned vs zamazenta-crowned-shield
* pikachu-starter vs pikachu-partner
* eevee-starter vs eevee-partner

That's fine, we don't have to resolve those right now. Just thought it was worth noting.

Also, note to self: i'll need to copy a few sprites to accommodate the new zygarde forms.

Parnassius commented on pokedex/data/csv/pokemon_egg_groups.csv:
```diff
@@ -357,8 +357,11 @@ species_id,egg_group_id
 278,4
 279,2
 279,4
+280,8
 280,11
+281,8
 281,11
+282,8
 282,11
```
apparently they weren't, they gained it in gen8

Parnassius commented on pokedex/data/csv/pokemon_evolution.csv:
```diff
@@ -358,7 +358,7 @@ id,evolved_species_id,evolution_trigger_id,trigger_item_id,minimum_level,gender_
 359,715,1,,48,,,,,,,,,,,,,,0,0
 360,678,1,,25,,,,,,,,,,,,,,0,0
-361,700,1,,,,,,,,18,,,2,,,,,0,0
+361,700,1,,,,,,,,18,160,,2,,,,,0,0
```
Whoops, i thought i did. Fixed

Parnassius commented on pokedex/data/csv/items.csv:
```diff
@@ -953,3 +953,135 @@ id,identifier,category_id,cost,fling_power,fling_effect_id
+1018,silver-pinap-berry,,1000,10,
+1019,golden-pinap-berry,,5000,10,
+1020,secret-key,,0,,
+1021,ss-ticket,,0,,
```
I added a `--letsgo` suffix, mainly not to interfere with item_game_indices.csv, since the game data seems to have both the new and the original items

Parnassius commented:
> Oh, one thing i missed from Let's Go - it looks like the partner forms of Pikachu and Eevee are missing. They probably got lost when you dropped unobtainable pokemon from the rip.

Whoops, you're right, i dropped them when i removed the unobtainable pokemon, along with alolan forms (mainly relevant for pokemon_moves.csv)

Parnassius commented on pokedex/data/csv/pokemon_form_names.csv:
```diff
@@ -1086,6 +1258,16 @@ pokemon_form_id,local_language_id,form_name,pokemon_name
+773,8,Tipo Normale,
+773,9,Type: Normal,
+773,11,タイプ：ノーマル,
+773,12,属性：一般,
```
I think Normal Arceus was the only one missing a name, all the others already had one

Parnassius commented:
I just noticed there were 4 more items with duplicate identifiers, i just fixed them.
- `pokemon-box` (from let's go, row 1007) was renamed to `pokemon-box-link` in sword/shield, even though the "real" `pokemon-box-link` is row 1139. I reverted the identifier to the old one
- `leaf-letter`s are now called `leaf-letter--pikachu` and `leaf-letter--eevee`
- `fishing-rod` from sword/shield is now called `fishing-rod--galar`. The other one still refers to Lana from sun/moon in its description, so i would keep them separate
- `reins-of-unity`, i used the same `--merge` and `--split` suffixes already used for kyurem's and necrozma's items. In this case there's an additional third item (row 1658), which i left as-is. According to the description it seems to be the item you get in the dlc before you catch calyrex; when it's caught i think it becomes `reins-of-unity--split`, and after that upon splitting/fusing it switches between that one and `reins-of-unity--merge`, as expected

magical commented:
Looking good, thanks.

I tried loading the data into PostgreSQL and ran into a couple problems:
1. items.category_id is required
2. pokemon.shape_id can't be 0

Shapes: i'd be fine just marking this nullable for now. I *know* shapes still exist, at least in HOME, and we'll definitely want to rip them at some point, but there's no reason we have to do it right now.

Looks like was wrong about item categories, oops. It's true that categories are non-official and we don't strictly need them, but we *do* need to know the item pocket, and pockets are linked to categories. Pockets are semi-official: the current ones are from the D/P/Pt days and correspond roughly to the pockets in the bag from that game, although obviously lots of other games have come out since then that had slightly different pocket schemes. You can make a "sw/sh uncatagorized" category in one of the pockets and throw all the items in there for now. (If you want to try ripping them though, i think the pockets roughly correspond to [this field](https://github.com/Parnassius/i-pokemoni/blob/011808473ba3da1eb7555af73f4a6e6986c7a08a/item/item_info_swsh.py#L125) in the item data.)

---

A couple other things i noticed:

* TM25 is listed as Max Guard, but Bulbapedia says it is Protect. Is that right?
* Xerneas has no default form. `xerneas-neutral` should be the default.

Parnassius commented:
>     * TM25 is listed as Max Guard, but Bulbapedia says it is Protect. Is that right?

Whoops, fixed. I compared the list with both Bulbapedia's and Serebii's, and now everything seems to be ok

>     * Xerneas has no default form. `xerneas-neutral` should be the default.

[I'm dumb](https://github.com/Parnassius/i-pokemoni/commit/007982cb8b888151c6e8e6b89ca56c1d6c290ebd), fixed

> Shapes: i'd be fine just marking this nullable for now. I _know_ shapes still exist, at least in HOME, and we'll definitely want to rip them at some point, but there's no reason we have to do it right now.

The search filter is indeed missing in Sword and Shield's pokedex, I'm honestly not sure if the data is still available somewhere in the game files. Should I simply set `nullable=True` in [this line](https://github.com/veekun/pokedex/blob/a91a7d95b32915751985c8ddac25ff7211c3809e/pokedex/db/tables.py#L2038), or there's something more involved to do?

> Looks like was wrong about item categories, oops. It's true that categories are non-official and we don't strictly need them, but we _do_ need to know the item pocket, and pockets are linked to categories. Pockets are semi-official: the current ones are from the D/P/Pt days and correspond roughly to the pockets in the bag from that game, although obviously lots of other games have come out since then that had slightly different pocket schemes. You can make a "sw/sh uncatagorized" category in one of the pockets and throw all the items in there for now. (If you want to try ripping them though, i think the pockets roughly correspond to [this field](https://github.com/Parnassius/i-pokemoni/blob/011808473ba3da1eb7555af73f4a6e6986c7a08a/item/item_info_swsh.py#L125) in the item data.)

I had a couple of problems trying to rip them, since as you said the pockets are from D/P/Pt. These are the main things I'm unsure about:
- the let's go main menu buttons (pokemon-box/medicine-pocket/candy-jar/etc): I put them in key => unused, even though they are not properly key items
- [the various things eevee and pikachu can pick up](https://github.com/Parnassius/i-pokemoni/blob/0b7c3ca37c44608c40ec8bea474e4c321d654c7b/item/item_info_letsgo.py#L103) in let's go: I put them in misc => loot, I don't think there's anything better?

I also created a couple new categories:
- misc => species-candies
- berries => catching-bonus (golden/silver razz/nanab/pinap berries)
- misc => dynamax-crystals
- medicine => nature-mints
- misc => curry-ingredients

Everything else seemed to be pretty straightforward

(I squashed almost everything together because I can't git apparently, if for any reason you need to check the individual commits they're still available [here](https://github.com/Parnassius/pokedex/commits/letsgo-swsh-3))

magical commented:
Thanks. I'll take another look in case i missed something, but i want to try and get this merged this weekend.

> The search filter is indeed missing in Sword and Shield's pokedex, I'm honestly not sure if the data is still available somewhere in the game files.  Should I simply set nullable=True in this line, or there's something more involved to do?

np, I'll take care of it.

magical commented:
Merged. Thank you!
